File: Blackbaud.Interview.Cards/Deck.cs
namespace Blackbaud.Interview.Cards;

/// <summary>
/// A deck of cards
/// </summary>
public class Deck
{
    private readonly Stack<Card> _stackOfCards;

    /// <summary>
    /// Private constructor for a new deck of <paramref name="cards"/>.
    /// Use Deck.NewDeck() static factory method.
    /// </summary>
    /// <param name="cards"></param>
    private Deck(IEnumerable<Card> cards)
    {
        _stackOfCards = new Stack<Card>(cards);
    }

    /// <summary>
    /// Creates and returns a new deck of cards.
    /// </summary>
    /// <returns></returns>
    public static Deck NewDeck()
    {
        return new Deck(
            Enum.GetValues<Suit>().SelectMany(suit =>
                Enum.GetValues<Rank>().Select(rank =>
                    new Card(rank, suit))
        ));
    }

    /// <summary>
    /// The number of remaining cards in the deck
    /// </summary>
    public int RemainingCards => _stackOfCards.Count;

    /// <summary>
    /// Returns true if there are no remaining cards in the deck
    /// </summary>
    public bool Empty => RemainingCards == 0;

    /// <summary>
    /// Removes the next card from the deck.
    /// </summary>
    /// <returns>The next card from the deck.
    /// Returns null if no cards remain.</returns>
    public Card NextCard()
    {
        if (!Empty)
        {
            var nextCard = _stackOfCards.Pop();
            return nextCard;
        }
        else
        {
            return null;
        }
    }

    // Shuffle the deck
    public void Shuffle()
    {
        var cards = _stackOfCards.ToArray();
        var random = new Random();
        for (int i = cards.Length - 1; i > 0; i--)
        {
            int n = random.Next(i + 1);
            (cards[n], cards[i]) = (cards[i], cards[n]);
        }
        _stackOfCards.Clear();
        foreach (var card in cards)
        {
            _stackOfCards.Push(card);
        }
    }
}

File: Java/src/Deck.java
import java.util.Arrays;
import java.util.Collection;
import java.util.List;
import java.util.Stack;
import java.util.stream.Collectors;

public class Deck {
    private Stack<Card> cards;

    private Deck(Collection<Card> cards) {
        this.cards = new Stack<>();
        this.cards.addAll(cards);
    }

    public static Deck newDeck() {
        List<Card> cards = Arrays.stream(Suit.values())
            .flatMap(suit ->
                Arrays.stream(Rank.values())
                    .map(rank -> new Card(rank, suit))
            ).collect(Collectors.toList());
        return new Deck(cards);
    }

    public int countRemainingCards() {
        return cards.size();
    }

    public boolean isEmpty() {
        return cards.isEmpty();
    }

    public Card nextCard() {
        if (isEmpty() == false) {
            return cards.pop();
        } else {
            return null;
        }
    }
}
